core/util/http.py

```diff
@@ -430,52 +434,55 @@ def debuggable_request(
-        allowed_response_codes=None,
-        expected_encoding="utf-8",
-    ):
+        response: Response,
+        allowed_response_codes: Optional[List[Union[str, int]]] = None,
+        disallowed_response_codes: Optional[List[Union[str, int]]] = None,
+        expected_encoding: str = "utf-8",
+    ) -> Response:
         """If there was a problem with an integration request,
-        return an appropriate ProblemDetail. Otherwise, return the
+        raise ProblemError with an appropriate ProblemDetail. Otherwise, return the
         response to the original request.
 
         :param response: A Response object from the requests library.
-        :param expected_encoding: Typically we expect HTTP responses to be UTF-8
+        :param expected_encoding: Typically, we expect HTTP responses to be UTF-8
             encoded, but for certain requests we can change the encoding type.
         """
 
         allowed_response_codes = allowed_response_codes or ["2xx", "3xx"]
-        allowed_response_codes = list(map(str, allowed_response_codes))
+        allowed_response_codes_str = list(map(str, allowed_response_codes))
+        disallowed_response_codes = disallowed_response_codes or []
+        disallowed_response_codes_str = list(map(str, disallowed_response_codes))
+
         code = response.status_code
         series = cls.series(code)
-        if str(code) in allowed_response_codes or series in allowed_response_codes:
-            # Whether or not it looks like there's been a problem,
+        if (
+            str(code) in allowed_response_codes_str
+            or series in allowed_response_codes_str
+        ):
+            # Whether it looks like there's been a problem,
             # we've been told to let this response code through.
             return response
 
         content_type = response.headers.get("Content-Type")
-        response_content = response.content
-        if response_content and isinstance(response_content, bytes):
-            try:
-                response_content = response_content.decode(expected_encoding)
-            except Exception as e:
-                return RequestNetworkException(url, e)
         if content_type == PROBLEM_DETAIL_JSON_MEDIA_TYPE:
             # The server returned a problem detail document. Wrap it
             # in a new document that represents the integration
             # failure.
-            problem = INTEGRATION_ERROR.detailed(
-                _("Remote service returned a problem detail document: %r")
-                % (response_content)
-            )
-            problem.debug_message = response_content
-            return problem
+            try:
+                problem_detail = INTEGRATION_ERROR.detailed(
+                    f"Remote service returned a problem detail document: '{response.text}'"
+                )
+                problem_detail.debug_message = response.text
+                raise ProblemError(problem_detail=problem_detail)
```

Comment:
This class is used pretty much everywhere, but I don't think all the places where it is used handle ProblemErrors. Would this be problematic?

Reply:
The HTTP class is used pretty much everywhere, but from what I can tell, the function modified here to raise a `ProblemError` `process_debuggable_response` is only called from `debuggable_get` and `debuggable_post`. And those functions are only called in the `OpdsRegistrationService` code, which does handle the resulting `ProblemError`.

So I don't think these changes will be problematic, but let me know if you see uses of the `debuggable_*` functions I missed.